### ISSUE TITLE
Node bound entities

### ISSUE THREAD

**Issue by bell07** [minetest/minetest#6963]
We have some mods that uses entities to take additions for nodes because of static nature of node. It is legit usage. But the mods have all issues to get the node and the entity in sync. The most implementations uses minetest.get_objects_inside_radius(pos, 0.75) to get the connection between node and entity that is inaccurate.

Feature request: Node bound entities

We need a way to assign an entity to a node. That means:
- The entity is loaded if the node gets active/loaded
- The entity is not destroyed by clearobjects
- The entity is destroyed if node is removed
- The entity is created if node is placed (new callback?)

I do not know in detail how the request could be implemented in detail, but I think it is lucidly what I mean. A solution to avoid https://github.com/minetest-mods/signs_lib/issues/24 and other similar issues like entity exists after node was removed or entity missed for unknown reason.

**Comment by Wuzzy2** [comment]
This is a very good idea, I have thought of this for my subgame as well. This will help me fix a lot of problems.
Especially the fact that `clearobjects` agressively deletes *every* entity is a big problem.

I think the idea here is to make a simplified and intentionally restricted entity. This will be useful.

I think one core property of a node-bound entity could be that it does not move (i.e. changes position, but an animation is OK). This could have some good implications for server performance.

Node-bound entities also should be rendered earlier than other entities.

Another important thing of node-bound entities is that they support animations flawlessly. The goal should be that chest open/close animations should work perfectly without 100000 lines of scripting overhead.

Use cases:
- Fluid chest open/close animation
- Text on signs
- Item in item frame or on pedestal
- Animated nodes

**Comment by raymoo** [comment]
>You can store the entity reference object in the table of another entity, if node meta could store tables that could work for the base problem. Otherwise you could use the entity identifier iirc.

A reference object can be come invalid, or worse, can come to refer to another entity entirely if the ID is reused. I'm not sure what an entity identifier is, or whether it's exposed to Lua, but I agree that some kind of identifier system would be useful, along with an entity equivalent of `get_player_by_name`.

**Comment by stujones11** [comment]
Rather than binding entities to specific nodes, I would prefer to see a new type of object altogether. Something like a static entity that could support armature animation and dynamic texturing but otherwise behaves like a regular node.

**Comment by raymoo** [comment]
@stujones11 Like Minecraft tile entities?

**Comment by stujones11** [comment]
No idea :P)

**Comment by raymoo** [comment]
@stujones11 I'm not sure exactly how they work, but every node that's animated in MC is something called a TileEntity.

I feel like with persistent IDs though, node entities could be implemented in just Lua. If transient entities are implemented then it would be even better.

**Comment by bell07** [comment]
I think an important behaviour change for node entities is the entities should not be loaded like usual entities but re-created each time the node is loaded.
The issue I had with signs_lib was I deactivated a mod, the node becomes unknown_node but the entity was still at place. Of course it is possible to implement self-destroying the entity, but what happens if I re-enable the mod again?
Other simpler use case is if the signs_lib is enabled the first time, all existing signs needs to be punched for update. If entity is not loaded but re-created each node load, both issues cannot appear anymore.

An other connected idea is to be able to overwrite the node bound entities client site. That allow to use client-site-signs mod without anything on server or reversed - disable such entities client site on slow machines

**Comment by lhofhansl** [comment]
How accurate does this have to be? Is it good enough to get all objects for the mapblock (16x16x16 nodes)?
(I guess the answer is "no", but I wanted to double check)

**Comment by bell07** [comment]
The request is not to easy getting nearly objects but the bounding of entity to node. The request is to be able to assign multiple entities to a placed node and control them by node. The entity does not need own identity, should be managed by the node. Created on load or on place node, and removed on unload or on remove/swap node.

The node needs new function `on_load`, that is nearly the LBM calls. In this function the implementation could be done for the objects using
`minetest.add_node_bound_entity(nodepos, entitypos, "name", [staticdata])`

get_staticdata is not needed because the entity should not be stored. If the entity is unloaded it should be just removed. And if the node gets activated again, the entity is re-created in on_load

on_step is not needed. The node's timer can be used for dynamic updates if needed. At this point we need
`minetest.get_node_bound_entities(nodepos)` - returns a named list of node bound entities
`minetest.get_node_bound_entitity_byname(nodepos, name)` - Returns named entity


 Some other functions and attributes could be removed too, as proposed by Wuzzy2, this type of entities could be a "light" version.

**Comment by Ferk** [comment]
The thing is that I believe the range in which entities are active might be different than the range that nodes are active.

An LBM type of activation to recreate the entity might not work that well. If that worked you could already do that with an LBM and removing the entity when activated without data, or maybe you can already remove it from get_staticdata.

There would need to be a node callback for when the player gets into the entity activation range. But I don't know if then that might be too intensive for the server to check for lua callbacks in all nodes every time the player moves. Though if the entity is just for some visual effect, then maybe it could be just a special kind of entity that only exists client-side.

**Comment by bell07** [comment]
I do not think we need an own activation range for such entities. The Node's range as used in LBM's should be enough. The hook "on_load" should be after LBM-processing but before the timers and ABM's triggered. And of course in minetest.set_node() after on_construct, and before ABM's triggered.

If the entity is far away to the bounded node, it is bad programming. So the node will be unloaded if the player is away from node but far to the entity, or loaded even it is far but the node is nearly.

Of course it will be nice to have such bounded entities client-site. This allow to get more decorations on faster machines and skip them on slower machines (Open-chest-animation for example).

But we cannot force to install the client-site part of the mod if the serversite part requires it. I think about mods like [Item shelf](https://forum.minetest.net/viewtopic.php?f=9&t=19463) that cannot implemented client-site only. Btw. If I enable the mod server-site it is useless without the client-site support. How I can force the client to install the client-site part? I think we need it server-site at the first.

**Comment by Ferk** [comment]
> The Node's range as used in LBM's should be enough

The thing is that I don't think entities have that wide of a range. You can test this out in creative. Drop a lot of items and fly away enough that they disappear, even though the nodes are still there.
Then quickly fly back and if you are fast enough you'll see there's a small delay until the entities re-activate, you can see them pop up.

Since the entity deactivated without the node really "unloading" the "on_load" wouldn't have been called so an entity getting removed wouldn't reappear properly. Also the server has a limit on how many entities can be active at the same time, client side. It's not only the distance that's determining if it'll be unloaded.

So these node entities would either have to be special to not deactivate only if the node is deactivated or the callback of the node would have to be based on entity (de)activation and not LBM loading.

If LBM loading is really enough for you, what's stopping you from just using current LBMs?

**Comment by bell07** [comment]
I know the usual entities does have an other range. An important part of this request is to create new type of entities with lifecycle bound to a node, and not an own lifecycle depending on entity ranges. The new entity object (NBSAO ? = Node Bound Server Active Object?) should not be stored in the world and cannot be destroyed by clearobjects. The NBSAO does less functionality then the usual SAO to be "light".

The reqested range for NBSAO: I do not know the nodes implementation in detail, but if you go away from a node at some distance the ABM's and the "on_timer" are not called anymore. This point I mean with "unloaded".  This is the right time to call "nodedef.on_unload" and destroy the NBSAO. If you come back, the ABM's and on_timer are called. Before this both the node's "nodedef.on_load" should be called to re-create the NBSAO's.

The LBM using run_at_every_load is not enough because the minetest.set_node and minetest.swap_node needs to be covered too. If a node is replaced the old node needs to be checked for NBSAO and destroy them, and the new node needs to call the on_load to create new NBSAO if supported. In best case it should work if an area is changed using VoxelManip object.

**Comment by ZNixian** [comment]
I'd be interested in implementing this. I agree with @bell07 about the load/unload system. Ideally, I think this should be kept completely separate from the standard entity system.

One other thing should be a `get_nbsao_at_pos` function, or something.

One other thing about this: it would be nice to be able to insert some kind of element into a formspec, that reflects the result of a function in the NBASO, updated every step if it's value has changed. This would allow for formspecs with smoothly changing values - such as the charge stored in some electronic device.

**Comment by bell07** [comment]
> One other thing about this: it would be nice to be able to insert some kind of element into a formspec, that reflects the result of a function in the NBASO, updated every step if it's value has changed. This would allow for formspecs with smoothly changing values - such as the charge stored in some electronic device.

Good point, but I think the formspec is not the right place for this. Beside to "formspec" and "infotext" maybe an additional metadata (nbsao_staticdata?) coupd be implemented that triggers nbsao update on change. So it is possible to update the nbsao from on_receive_fields in a easy way by `meta:set_string("nbsao_staticdata",value)`

Could be server-site only (trough mark_as_private) or for both, client- and server

**Comment by ZNixian** [comment]
> Good point, but I think the formspec is not the right place for this. Beside to "formspec" and "infotext" maybe an additional metadata (nbsao_staticdata?) coupd be implemented that triggers nbsao update on change. So it is possible to update the nbsao from on_receive_fields in a easy way by
>
> `meta:set_string("nbsao_staticdata",value)`

I was thinking that NBSAOs would update every step, rather than being manually triggered (though this could be added as an option when registering the NBSAO, probably as a value passed to `minetest.register_node`).

I was more thinking of a way to sync a constantly updating value with clients. If I correctly interpreted what you said, you thought I meant as a way to react when a formspec button was clicked?

**Comment by bell07** [comment]
I retract the nbsao_staticdata proposal, forgotten that we should be flexible to support multiple NBSAO's per node (itemshelf mod use case). So instead of get_nbsao_at_pos(pos) I proposed already 
```
minetest.get_node_bound_entities(nodepos) - returns a named list of node bound entities
minetest.get_node_bound_entitity_byname(nodepos, name) - Returns named entity
```
functions could be renamed, but we need to support a list. 

I dislike the updating every step because the NBSAO should be light. If updates needed every step the nodetimer could be utilized or an ABM to update all entites at once for the position.

> If I correctly interpreted what you said, you thought I meant as a way to react when a formspec button was clicked?

Yes and no. I try to explain on an example: https://github.com/Gerold55/minetest-laptop/issues/34
Laptop does simplate working OS trough node-formspec. The requirement is to show the screen formspec on the monitor texture. An NBSAO could be used for this. Because we do not have a formspec2texture renderer in first step the selected theme-background image should be displayed on screen.
The formspec is a static thing, any change is triggered by lua-runtime. That could be from on_receive_fields, from timer or from any other lua-code-peace. 

In case formspec2texture is available it is possible to implement the synchronization between the formspec and the NBSAO at the same place as formspec is updated, like

```
meta:set_string('formspec', formspec)
minetest.get_node_bound_entitity_byname(node.pos, 'display'):set_properties(texture=formspec2texture (formspec))
```
Without formspec2texture I can do it just on theme-switch.
```
os_class:set_theme(theme)
....
minetest.get_node_bound_entitity_byname(node.pos, 'display'):set_properties(texture=self.theme.launcher_bg)
```

What is your use-case? If I understand you right you like to see the formspec2texture integrated in NBSAO? I think it should be an other future-request, because is usable for "normal" entities or on other formspecs (preview) too.

**Comment by ZNixian** [comment]
> I retract the nbsao_staticdata proposal, forgotten that we should be flexible to support multiple NBSAO's per node (itemshelf mod use case). So instead of get_nbsao_at_pos(pos) I proposed already

This sounds somewhat painful to implement. I was imagining defining the the NBSAO when you called `minetest.register_node`, and if the mod needs multiple, it could handle that itself (though come to think about it, it wouldn't be any good for 3d models).

> I dislike the updating every step because the NBSAO should be light. If updates needed every step the nodetimer could be utilized or an ABM to update all entites at once for the position.

Maybe don't update by default, but include an option to do so. The use case I was imagining was high-speed electricity-through-cable propagation, so electricity wouldn't pulse though every `x` seconds, but rather transfer constantly.

In this case, it could store data in it's variables, that could then be written to meta when unloaded, to avoid updating the database every frame. Assuming NBSAOs were somehow marked as invalid if their associated block were to be broken, you could fetch the object for another block then read and write it's variables with negligible performance impact, even when doing so many times per frame.

**Comment by bell07** [comment]
My idea about multiple instances was something like
`minetest.register_nbsao(nbsao_name, node_name, nbsao_def)`
Of course it is possible to use allways 1x nbsao with 3d model, but for people like me they cannot create own models, using multiple I can just use 8x times the '__item' model on 8x nbsao definitions for itemshelf like mod.

About your prevous post: I do not understand why you asked for new formspec-element. 

The described requrement "high speed cable" is basically like mesecons that is already implemented without NBSAO.  Some visuals could be improved trough NBSAO. The "high-speed electricity-through-cable propagation" visual could be done trough animation on NBSAO, and animation speed adjustment if the value is changed (nighboar block broken for example) so no permanent updates still needed.

The NBSAO should not have own data and own calls, It is an node enhancement, so all data should be in nodemeta and all functions should be controlled from node's function. For this we need an "on_enable" and "on_disable" node functions to right init/de-init the (multiple) NBSAO

**Comment by ZNixian** [comment]
> My idea about multiple instances was something like
> `minetest.register_nbsao(nbsao_name, node_name, nbsao_def)`

So you'd manually add NBASOs to a node after placing it in the world, rather than it being created automatically?

> Of course it is possible to use allways 1x nbsao with 3d model, but for people like me they cannot create own models, using multiple I can just use 8x times the '__item' model on 8x nbsao definitions for itemshelf like mod.

Gotcha

> About your prevous post: I do not understand why you asked for new formspec-element.

You'd be able to enter something into a formspec, which would show a constantly updating number.

```
dyn_label[X,Y;id]
```

```
...
get_label_contents = function(self, id)
    return "hi: " .. tostring(self.time_since_block_placed)
end
...
```

This would show the time since the block was placed, and the number would continue to grow every step while the formspec was open.

> The described requrement "high speed cable" is basically like mesecons that is already implemented without NBSAO. Some visuals could be improved trough NBSAO. The "high-speed electricity-through-cable propagation" visual could be done trough animation on NBSAO, and animation speed adjustment if the value is changed (nighboar block broken for example) so no permanent updates still needed.

Think Technic or Industrialtest, where energy is a limited supply. I'm not concerned about graphics, however.

> The NBSAO should not have own data and own calls, It is an node enhancement, so all data should be in nodemeta and all functions should be controlled from node's function.

I was thinking of using it for caching data that needs to be accessed (including being written to) potentially hundreds of times per step, which I'm fairly certain would cause performance issues with nodemeta. However, nodemeta would be the only permanent way to store stuff.

> For this we need an "on_enable" and "on_disable" node functions to right init/de-init the (multiple) NBSAO

I agree.

**Comment by bell07** [comment]
> So you'd manually add NBASOs to a node after placing it in the world, rather than it being created automatically?
That was the initial idea but I know it is not the best one. I like to get the "Gotcha". Do you see an other way to support multiple NBSAO per node?

About formspec: Assume get_label_contents() and self.time_since_block_placed calculation should be called each globalstep? On server and client? I think the formspec optimizations should be not a part of this issue, because no direct dependency.

> I was thinking of using it for caching data that needs to be accessed (including being written to) potentially hundreds of times per step

If you are not about graphics, you just need the  "on_enable" / "on_disable" hooks on node definition. If both methods exists you can just use own table to handle the cache. Should be faster then the full NBSAO framework

```
local mycache = {}
register_node(mynode, {
...
on_enable = function(node, pos)
    mycache[minetest.hash_node_position] =  minetest.deserialize(minetest.get_meta(pos):get_string("data")) or {}
end,
on_disable = function(node, pos)
    minetest.get_meta(pos):set_string("data", minetest.serialize(mycache[minetest.hash_node_position]))
    mycache[minetest.hash_node_position] = nil
end,
})
```

**Comment by Ferk** [comment]
You are gonna be writing to the metadata on every node unload?

I still think it would be better to have the entities be completely different, and not be SAOs so that they don't need to stored or managed server-side and avoid any load/unload problems. Managing things through callbacks makes it complicated and prone to problems.

They basically would be more like animated meshes for decoration purposes and less like actual entities.

**Comment by bell07** [comment]
>You are gonna be writing to the metadata on every node unload?

That was my proposal for the
>However, nodemeta would be the only permanent way to store stuff.

>I still think it would be better to have the entities be completely different, and not be SAOs so that they don't need to stored or managed server-side and avoid any load/unload problems. Managing things through callbacks makes it complicated and prone to problems.

Agree the entities should be complete different and more lighten then SAP. No hooks, no callbacks, just basic attributes like texture, model and position. All other shold be done from Node. The both hooks on_load and on_unload should be on node, not on the entity.
The entity is created automatically before the on_load (or manually inside the on_load) and destroyed after on_unload. 

Do you plan to implement it client-site only (NPCAO?). I think at least the mentioned basic attributes should be accessable/changeable at runtime serversite.

**Comment by bell07** [comment]
> If the entity is just for some visual effect, then maybe it could be just a special kind of static entity that only exists client-side and that can be always active as long as the node is there. Then have the data for the entity creation in, for example, a field from the metadata of the node sort of like the "infotext", for it to be loaded client-side.

I read it again and may this could be an solution. Writing some attirbutes (position, texture, model, animation, animation speed, ..???) to the nodemeta that is enough to draw the NPCAO on client site only. Maybe in a way multiple instances possible per node? On serversite no enhancements needed, the mods does just write to this nodemeta field.
At the client site if data is received about the node including the metadata, the NBCAO is created. If An update appears for this node, the assigned NBCAO is updated or removed. Maybe some hooks to manipulate the NBCAO trough CSM.
That matches my initial issue. The on_load and on_unload could be an other future request

**Comment by ZNixian** [comment]
Good point about being able to use a `on_load` and `on_unload` hook, that'd work fine. One problem with putting them into meta values is animations, particularly if you wanted the animations to match between clients.

**Comment by bell07** [comment]
Yes, the animation synchronization is needed. Not only to match between clients but more important to match between nodes on the same client but loaded at different times. I think additional meta "animation timestamp" should be added/transferred so each client can proper calculate the current animation state and sync against this time. Timestamp should be clearly, previous synchronized, eather game time or server time.

**Comment by numberZero** [comment]
I consider this feature really important. It has tons of uses which we may or may not imagine. Text on signs is just a single example. Armor stand is a better example, as it behaves like a node in some aspects but more like an entity in other ones. Animated chests are nice example as well: nodes are expected to be static normally, but interaction may and should be animated well.

There was a suggestion of mostly-client-side implementation. While it would be basically suitable for most imaginable use cases, it requires storing all the entity data permanently in meta. That’s not always good, and often makes keeping all the data in sync harder.

**Comment by raymoo** [comment]
Pairing nodes and entities would be possible to implement in lua as a mod if it wasn't for entity count limits (including small ID size) and performance issues.

**Comment by bell07** [comment]
They are a lot of lua implementations, but all of them are hacky and leads to different issues, because the requested functionalitiy is missed.

> The entity is loaded if the node gets active/loaded
> The entity is not destroyed by clearobjects
> The entity is destroyed if node is removed
> The entity is created if node is placed (new callback?)

**Comment by raymoo** [comment]
>The entity is loaded if the node gets active/loaded

Could be done with a spatial data structure and checking around every player in a global step function. Keep a mapping from node positions to entity data.

>The entity is not destroyed by clearobjects

True that it can't be truly prevented, but if the entity only uses node meta to store its data (no storing eventually-persistent data in the luaentity), then it could be restored without data loss by a suitably-implemented system. Maybe get_staticdata could be called in a globalstep callback but I don't know if it would be timed correctly to behave completely correct.

>The entity is destroyed if node is removed

This one is easy, just have the entity remove itself if its node is gone and remove any bookkeeping data for the pairing

>The entity is created if node is placed (new callback?)

Similar to above, create the entity and bookkeeping info in on_construct or something

**Comment by numberZero** [comment]
@raymoo The whole point of this issue is to add an *lightweight* node-bound entity for various purposes, from text display to animation.

> Changing animation can be done by metadata update for example.

Please **don’t.**

Actually animation could be transferred as Lua code to be executed in an isolated environment. Or more conventionally, as a key frame sequence.

**Comment by bell07** [comment]
>> Changing animation can be done by metadata update for example.

> Please don’t.

I did not meant the animation is transferred, but controlled, like
```
minetest.get_meta(pos):set_string("node_entity", minetest.serialize({
	mesh = "the_mesh.obj",
	animation = { {x=1,y=1}, frame_speed=15, frame_blend=0 },
	textures = { "the_texture.png" },
)
```
The client should have the mesh and know how the animation parameters should be prosessed

**Comment by paramat** [comment]
1 year bump. Any core dev support?

**Comment by paramat** [comment]
:+1:

**Comment by appgurueu** [comment]
Probably it should just be an entity *attached* to the node ? Like overloading attaching ? Entity would remain attached even if the node fell...

**Comment by AbbyRead** [comment]
I don't know how you guys feel about low-purpose comments, but I would love this feature.  I've been working on a game recently and just thought this was already a thing.  Nbd I guess if it's simple to add an entity on node load, but it does feel a bit strange and extra to have to do it that way.

**Comment by Ghurir** [comment]
if someone is trying to figure out how to do this, currently you have to do something like the following:
### node
```
local node_def = {
    on_construct = function(pos)
        local meta = core.get_meta(pos)
        mod.overlay_each_face(pos, meta)
    end,

    after_dig_node = function(pos, oldnode, oldmeta, digger)
        mod.remove_overlays(pos)
    end,
}

-- when you get too far entity unloads by itself, so you need to load it with the node
core.register_lbm({
    label = "Load block entity",

    name = "mod:load_block",

    nodenames = {"mod:block", "mod:block_warm"},
    -- Names of non-registered nodes and groups (as group:groupname)
    -- will work as well.

    run_at_every_load = true,
    -- If `true`: The LBM runs every time a mapblock is activated.

    bulk_action = function(pos_list, dtime_s)-- (available since 5.10.0)
        for i,pos in ipairs(pos_list) do
            local meta = core.get_meta(pos)
            mod.overlay_each_face(pos, meta)
        end
    end,
})
```
### entity
```
core.register_entity("mod:overlay", {
    initial_properties = {
        visual = "upright_sprite",
        textures = {"blank.png"},
        physical = false,
        pointable = false,
    },

    on_activate = function(self, staticdata)
        local data = core.deserialize(staticdata)
        -- so we can set the texture when spawning the entity
        if data then
            self.object:set_properties({
                textures = {data.texture}
            })
        end
    end
})
```
### api
```
function mod.spawn_overlay(pos, texture, direction)
    local o = offsets[direction]
    -- the texture is set when spawning entity
    local obj = core.add_entity(
        vector.offset(pos, o.x, o.y, o.z),
        "mod:overlay",
        core.serialize({
            texture = texture
        })
    )
    obj:set_rotation(vector.new(o.x, o.y, o.z):dir_to_rotation())

    return obj
end

function mod.remove_overlays(pos)
    local objs = core.get_objects_inside_radius(pos, 1.0)
    for _, obj in ipairs(objs) do
        local entity = obj:get_luaentity()
        if entity and entity.name == "mod:overlay" then
            obj:remove()
        end
    end

end

function mod.overlay_each_face(pos, meta)
    for key, dir in pairs(mod.get_valid_directions()) do
        local name= meta:get_string(dir .. "_face")
        local tex = textures[name]
        if tex ~= nil then
            mod.spawn_overlay(pos, tex, dir)
        end
    end
end
```
